src/riverford/durable_ref/scheme/redis/carmine.clj

```diff
@@ -0,0 +1,152 @@
+        path (.getPath sub-uri)]
+    (if-let [connection (get @connections (.getHost sub-uri))]
+      connection
+      (let [new-connection {:pool {}
+                            :spec {:host (.getHost sub-uri)
+                                   :port (let [port (.getPort sub-uri)]
+                                           (if (pos? port)
+                                             port
+                                             6379))}}]
+        (swap! connections assoc (str (get-in new-connection [:spec :host]) ":" (get-in new-connection [:spec :port])) new-connection)
+        new-connection))))
+
+(defn- location
+  [^URI uri]
+  (let [sub-uri (URI. (.getSchemeSpecificPart uri))
+        path (.getPath sub-uri)
+        [_ db id-or-ns ?id] (str/split path #"/")]
+    [(edn/read-string db)
+     (when ?id id-or-ns)
+     (or ?id id-or-ns)]))
+
+(defn- get-credentials
+  [opts connection]
+  (let [c1 (:credentials opts)
+        c2 (get @credentials
+                (str (get-in connection [:spec :host])
+                     ":"
+                     (get-in connection [:spec :port])))]
+    (or c1 c2)))
+
+(defmethod dref/write-bytes! "redis"
+  [^URI uri bytes opts]
+  (let [connection (uri->connection uri)
+        credentials (get-credentials opts connection)
+        [db ?f k] (location uri)]
+    (car/wcar (-> connection
+                  (assoc-in [:spec :password] (or (:password credentials) ""))
+                  (assoc-in [:spec :db] db))
+              (if ?f
+                (car/hset ?f k bytes)
+                (car/set k bytes)))))
+
+(defmethod dref/read-bytes "redis"
+  [^URI uri opts]
+  (let [connection (uri->connection uri)
+        credentials (get-credentials opts connection)
+        [db ?f k] (location uri)
+        conn-w-creds (-> connection
+                         (assoc-in [:spec :password] (or (:password credentials) ""))
+                         (assoc-in [:spec :db] db))]
+    (car/wcar conn-w-creds
+              (if ?f
+                (car/hget ?f k)
+                (car/get k)))))
+
+(defmethod dref/delete-bytes! "redis"
+  [^URI uri opts]
+  (let [connection (uri->connection uri)
+        credentials (get-credentials opts connection)
+        [db ?f k] (location uri)]
+    (car/wcar (-> connection
+                  (assoc-in [:spec :password] (or (:password credentials) ""))
+                  (assoc-in [:spec :db] db))
+              (if ?f
+                (car/hdel ?f k)
+                (car/del k)))))
+
+(defmethod dref/do-atomic-swap! "redis"
+  [^URI uri f opts]
+  (let [connection (uri->connection uri)
+        credentials (get-credentials opts connection)
+        [db ?f k] (location uri)
+        deserialize (dref/get-deserializer uri)
+        serialize (dref/get-serializer uri)
+        conn (-> connection
+                 (assoc-in [:spec :password] (or (:password credentials) ""))
+                 (assoc-in [:spec :db] db))]
+    (let [[_ [_ res-raw]] (car/atomic conn 100
```

Comment:
I think this will throw after 100 failed attempts under a contention scenario. I would prefer the default to be to keep trying in the event of CAS/optimistic lock failures, and to only throw on an unrelated exception (such as a connection issue). Clojure's atom will spin forever until it completes the swap, and I would like to carry that expectation over.  However, I do see the value in being able to supply an attempt limit as an opt - that would be cool.

Reply:
The problem here, I think, is that there can be a lot more pressure on a distributed atom than on a local one. Regardless whether it is a good idea or not to let it spin forever, this will need some more thinking and research on the Redis side of things; I have to figure out how to go about implementing the same functionality without the Carmine `atomic` macro.

I will start looking into this, but it will take some more time. :-)

Reply:
I think being able to provide spin policy (back-off config, limits, etc) via opts is a good idea - as long as the default is to spin forever so that IAtom meets expectations. Regardless of spin timeouts you want to be sure you are not using this lib in high contention scenarios as optimistic control like this never works well there. As for Redis I don't know I'm afraid would need to look into myself, I'll wait to see what you can come up with!